collect_app/src/main/java/org/odk/collect/android/http/okhttp/OkHttpOpenRosaServerClientFactory.java

```diff
@@ -0,0 +1,107 @@
+import okhttp3.Response;
+
+public class OkHttpOpenRosaServerClientFactory implements OpenRosaServerClientFactory {
+
+    private static final int CONNECTION_TIMEOUT = 30000;
+    private static final int WRITE_CONNECTION_TIMEOUT = 60000; // it can take up to 27 seconds to spin up an Aggregate
+    private static final int READ_CONNECTION_TIMEOUT = 60000; // it can take up to 27 seconds to spin up an Aggregate
+    private static final String USER_AGENT_HEADER = "User-Agent";
+    private static final String OPEN_ROSA_VERSION_HEADER = "X-OpenRosa-Version";
+    private static final String OPEN_ROSA_VERSION = "1.0";
+    private static final String DATE_HEADER = "Date";
+
+    private final OkHttpClient.Builder baseClient;
+    private final Clock clock;
+
+    public OkHttpOpenRosaServerClientFactory(@NonNull OkHttpClient.Builder baseClient, Clock clock) {
+        this.baseClient = baseClient;
+        this.clock = clock;
+    }
+
+    @Override
+    public OpenRosaServerClient create(String scheme, String userAgent, @Nullable HttpCredentialsInterface credentials) {
+        OkHttpClient.Builder builder = baseClient
+                .connectTimeout(CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                .writeTimeout(WRITE_CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                .readTimeout(READ_CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                .followRedirects(true);
+
+        if (credentials != null) {
+            Credentials cred = new Credentials(credentials.getUsername(), credentials.getPassword());
+
+            DispatchingAuthenticator.Builder daBuilder = new DispatchingAuthenticator.Builder();
+            daBuilder.with("digest", new DigestAuthenticator(cred));
+            if (scheme.equalsIgnoreCase("https")) {
+                daBuilder.with("basic", new BasicAuthenticator(cred));
+            }
+
+            DispatchingAuthenticator authenticator = daBuilder.build();
+
+            final Map<String, CachingAuthenticator> authCache = new ConcurrentHashMap<>();
+            builder.authenticator(new CachingAuthenticatorDecorator(authenticator, authCache))
+                    .addInterceptor(new AuthenticationCacheInterceptor(authCache)).build();
+        }
+
+        return new OkHttpOpenRosaServerClient(builder.build(), userAgent, clock);
+    }
+
+    private static class OkHttpOpenRosaServerClient implements OpenRosaServerClient {
+
+        private final OkHttpClient client;
+        private final String userAgent;
+        private final Clock clock;
+
+        OkHttpOpenRosaServerClient(OkHttpClient client, String userAgent, Clock clock) {
+            this.client = client;
+            this.userAgent = userAgent;
+            this.clock = clock;
+        }
+
+        @Override
+        public Response makeRequest(Request request) throws IOException {
+            return client.newCall(addHeaders(request)).execute();
+        }
+
+        private Request addHeaders(Request request) {
```

Comment:
How about making this static to make it really clear that it doesn't use anything from the instance? Maybe even calling it `withHeaders`? Or leaving it inline in `makeRequest`? For some reason I got tripped up for a second reading it.

Reply:
Right yeah I can totally see that. I think originally it was being called a few times but maybe now inlining is best. I'll give it a go.